GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs

```diff
@@ -110,18 +125,85 @@ public bool TryGetLooseObjectId(string directoryName, string filePath, out strin
         /// Creates a pack file from loose objects
         /// </summary>
         /// <returns>The number of loose objects added to the pack file</returns>
-        public int CreateLooseObjectsPackFile()
+        public CreatePackResult TryCreateLooseObjectsPackFile(out int objectsAddedToPack)
         {
-            int objectsAddedToPack = 0;
+            int localObjectCount = 0;
 
             GitProcess.Result result = this.RunGitCommand(
                 (process) => process.PackObjects(
                     "from-loose",
                     this.Context.Enlistment.GitObjectsRoot,
-                    (StreamWriter writer) => objectsAddedToPack = this.WriteLooseObjectIds(writer)),
+                    (StreamWriter writer) => localObjectCount = this.WriteLooseObjectIds(writer)),
                 nameof(GitProcess.PackObjects));
 
-            return objectsAddedToPack;
+            if (result.ExitCodeIsSuccess)
+            {
+                objectsAddedToPack = localObjectCount;
+                return CreatePackResult.Succeess;
+            }
+            else
+            {
+                objectsAddedToPack = 0;
+
+                if (result.Errors.Contains("is corrupt"))
+                {
+                    return CreatePackResult.CorruptBlob;
+                }
+
+                return CreatePackResult.UnknownFailure;
+            }
+        }
+
+        public string GetLooseObjectFileName(string objectId)
+        {
+            return Path.Combine(
+                            this.Context.Enlistment.GitObjectsRoot,
+                            objectId.Substring(0, 2),
+                            objectId.Substring(2, GVFSConstants.ShaStringLength - 2));
+        }
+
+        public int ClearCorruptLooseObjects(EventMetadata metadata)
+        {
+            int numBadObjects = 0;
+            int numTryGetIsValidFailures = 0;
+            int numFailedDeletes = 0;
+
+            foreach (string objectId in this.LooseObjectsBatch(this.MaxLooseObjectsInPack))
+            {
+                if (!this.Context.Repository.TryGetIsValidObject(objectId, out bool isBlob))
+                {
+                    numTryGetIsValidFailures++;
+                    continue;
+                }
+
+                if (!isBlob)
+                {
+                    string objectFile = this.GetLooseObjectFileName(objectId);
+
+                    if (this.Context.FileSystem.TryDeleteFile(objectFile))
+                    {
+                        numBadObjects++;
+                    }
+                    else
+                    {
+                        numFailedDeletes++;
+                    }
+                }
+            }
+
+            if (numTryGetIsValidFailures > 0)

```

Comment:
I would just put this in the `metadata` all the time even when zero.  Same with the `numFailedDeletes`.

Reply:
Agreed, especially since we only would add any metadata if something was corrupt.